File: backend/app/api/v1/contacts.py
from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Depends, Query, status
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session
import io
import csv

from app.core.database import get_db
from app.core.dependencies import get_workspace_id, require_crm_access
from app.models.user import User
from app.schemas.common import MessageResponse, PaginatedResponse
from app.schemas.crm.contact import ContactCreate, ContactResponse, ContactUpdate
from app.services.crm.contact_service import ContactService

router = APIRouter(prefix="/contacts", tags=["CRM - Contacts"])


@router.post("", response_model=ContactResponse, status_code=status.HTTP_201_CREATED)
def create_contact(
    data: ContactCreate,
    db: Session = Depends(get_db),
    company_id: UUID = Depends(get_workspace_id),
    current_user: User = Depends(require_crm_access),
):
    return ContactService(db, company_id).create(data, current_user.id)


@router.get("", response_model=PaginatedResponse[ContactResponse])
def list_contacts(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    account_id: Optional[UUID] = None,
    db: Session = Depends(get_db),
    company_id: UUID = Depends(get_workspace_id),
    _: User = Depends(require_crm_access),
):
    return ContactService(db, company_id).list_contacts(page, page_size, account_id)


@router.get("/export", response_class=StreamingResponse)
def export_contacts(
    db: Session = Depends(get_db),
    company_id: UUID = Depends(get_workspace_id),
    _: User = Depends(require_crm_access),
):
    contacts = ContactService(db, company_id).list_contacts(page=1, page_size=10000).items
    
    output = io.StringIO()
    writer = csv.writer(output)
    writer.writerow(["ID", "Name", "Email", "Phone", "Title", "Created At"])
    
    for contact in contacts:
        writer.writerow([
            str(contact.id),
            contact.full_name,
            contact.email or "",
            contact.phone or "",
            contact.job_title or "",
            contact.created_at.isoformat()
        ])
        
    output.seek(0)
    return StreamingResponse(
        iter([output.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": "attachment; filename=contacts.csv"}
    )

@router.get("/search", response_model=PaginatedResponse[ContactResponse])
def search_contacts(
    q: str = Query(..., min_length=1),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    company_id: UUID = Depends(get_workspace_id),
    _: User = Depends(require_crm_access),
):
    return ContactService(db, company_id).search(q, page, page_size)


@router.get("/{contact_id}", response_model=ContactResponse)
def get_contact(
    contact_id: UUID,
    db: Session = Depends(get_db),
    company_id: UUID = Depends(get_workspace_id),
    _: User = Depends(require_crm_access),
):
    return ContactService(db, company_id).get(contact_id)


@router.patch("/{contact_id}", response_model=ContactResponse)
def update_contact(
    contact_id: UUID,
    data: ContactUpdate,
    db: Session = Depends(get_db),
    company_id: UUID = Depends(get_workspace_id),
    _: User = Depends(require_crm_access),
):
    return ContactService(db, company_id).update(contact_id, data)


@router.delete("/{contact_id}", response_model=MessageResponse)
def delete_contact(
    contact_id: UUID,
    db: Session = Depends(get_db),
    company_id: UUID = Depends(get_workspace_id),
    _: User = Depends(require_crm_access),
):
    ContactService(db, company_id).delete(contact_id)
    return MessageResponse(message="Contact deleted successfully")

File: frontend/src/services/contacts.ts
import { api } from "@/lib/api";
import type { Contact, PaginatedResponse } from "@/types";

export interface ContactCreateData {
  first_name: string;
  last_name: string;
  email?: string;
  phone?: string;
  job_title?: string;
  contact_company?: string;
  account_id?: string;
  address?: string;
  notes?: string;
}

export const contactsService = {
  list(page = 1, page_size = 20): Promise<PaginatedResponse<Contact>> {
    return api
      .get<PaginatedResponse<Contact>>("/contacts", { params: { page, page_size } })
      .then((r) => r.data);
  },

  search(q: string, page = 1, page_size = 20): Promise<PaginatedResponse<Contact>> {
    return api
      .get<PaginatedResponse<Contact>>("/contacts/search", { params: { q, page, page_size } })
      .then((r) => r.data);
  },

  get(id: string): Promise<Contact> {
    return api.get<Contact>(`/contacts/${id}`).then((r) => r.data);
  },



  create(data: ContactCreateData): Promise<Contact> {
    return api.post<Contact>("/contacts", data).then((r) => r.data);
  },

  update(id: string, data: Partial<ContactCreateData>): Promise<Contact> {
    return api.patch<Contact>(`/contacts/${id}`, data).then((r) => r.data);
  },

  delete(id: string): Promise<void> {
    return api.delete(`/contacts/${id}`).then(() => undefined);
  },

  export(): Promise<Blob> {
    return api.get("/contacts/export", { responseType: "blob" }).then((r) => r.data);
  },
};
